File: python/377-Combination Sum IV.py
from typing import List
from functools import lru_cache

class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        "DFS"
        n = len(nums)
        @lru_cache(None)
        def dfs(total):
            if total > target:
                return 0
            if total == target:
                return 1
            result = 0
            for num in nums:
                result += dfs(total+num)
            return result
        return dfs(0)
    def combinationSum4(self, nums: List[int], target: int) -> int:
        "DP, Time: O(n*target), Space: O(target)"
        dp = [0]*(target+1) # dp[i] = numbers of combinations that can add up to i
        dp[0] = 1
        for i in range(1, target+1):
            for num in nums:
                if i - num < 0:
                    continue
                dp[i] += dp[i-num]
        return dp[-1]

nums = [1, 2, 3]
target = 4
result = Solution().combinationSum4(nums, target)
assert result == 7

File: cpp/377. Combination Sum IV.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

/*
dp[x] = sum(dp[x-num] for num in nums)
dp[0] = 1
ans = dp[target]
*/

class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        unsigned int dp[target+1]; fill(dp, dp+target+1, 0); dp[0] = 1;
        for(int x = 1; x <= target; ++x) {
            for(int num : nums) {
                if(x-num >= 0) dp[x] += dp[x-num];
            }
        }
        return dp[target];
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi nums; int target; int ans;

    solution = Solution();
    nums = {1, 2, 3};
    target = 4;
    ans = solution.combinationSum4(nums, target);
    assert(ans==7);
}